angular/projects/spark-angular/src/lib/components/sprk-tooltip/sprk-tooltip.component.ts

```diff
@@ -0,0 +1,237 @@
+import {
+  Component,
+  HostListener,
+  Input,
+  Output,
+  OnInit,
+  ViewChild,
+  EventEmitter
+} from '@angular/core';
+import { uniqueId } from 'lodash';
+
+@Component({
+  selector: 'sprk-tooltip',
+  template: `
+  <span class="sprk-c-Tooltip__container" #containerElement>
+    <button
+      class="sprk-c-Tooltip__trigger"
+      [ngClass]="{
+        'sprk-c-Tooltip__trigger' : true,
+        'sprk-c-Tooltip--toggled' : isOpen
+      }"
+      [attr.aria-expanded]="isOpen ? 'true' : 'false'"
+      [attr.data-analytics]="analyticsString"
+      [attr.data-id]="idString"
+      #triggerElement
+    >
+      <sprk-icon
+        [iconType]="triggerIconType"
+        [additionalClasses]="additionalClassesIcon"
+        aria-hidden="true"
+      ></sprk-icon>
+    </button>
+    <span
+      [ngClass]="getTooltipClasses()"
+      class="sprk-c-Tooltip"
+      aria-hidden="true"
+      role="tooltip"
+      #tooltipElement
+    >
+      <ng-content></ng-content>
+    </span>
+  </span>
+  `
+})
+export class SprkTooltipComponent implements OnInit {
+
+  /**
+   * The name of the icon to use
+   */
+  @Input()
+  triggerIconType: string;
+  /**
+   * The value supplied will be assigned to the
+   * `data-analytics` attribute on the component.
+   * Intended for an outside
+   * library to capture data.
+   */
+  @Input()
+  analyticsString: string;
+  /**
+   * Expects a space separated string
+   * of classes to be added to the
+   * component container.
+   */
+  @Input()
+  additionalClasses: string;
+  /**
+   * Expects a space separated string
+   * of classes to be added to the
+   * svg icon.
+   */
+  @Input()
+  additionalClassesIcon: string;
+  /**
+   * The value supplied will be assigned
+   * to the `data-id` attribute on the
+   * component. This is intended to be
+   * used as a selector for automated
+   * tools. This value should be unique
+   * per page.
+   */
+  @Input()
+  idString: string;
+  /**
+  * Optional: the unique ID to use for the tooltip element
+  */
+  @Input()
+  id: string = uniqueId(`sprk_tooltip_`);
+
+  /**
+   * Emitted when the tooltip is toggled open
+   */
+  @Output()
+  openedEvent = new EventEmitter<any>();
+
+  /**
+   * Emitted when the tooltip is toggled closed
+   */
+  @Output()
+  closedEvent = new EventEmitter<any>();
+
+  @ViewChild('triggerElement') triggerElement;
+  @ViewChild('tooltipElement') tooltipElement;
+  @ViewChild('containerElement') containerElement;
+
+  /**
+   * @ignore
+   */
+  @HostListener('document:keydown', ['$event'])
+  onKeydown($event) {
+    if ($event.key === 'Escape' || $event.key === 'Esc' || $event.keyCode === 27) {
```

Comment:
If we move the isOpen check before the checking of the key, it reduces the complexity. If the tooltip is not open then we don’t need to check what key it is at all.